File: frontend/src/components/sheet/widgets/formula.ts
/**
 * Formula grammar for computed sheet widgets — mirrors
 * backend/src/loregraph/schemas/formula.py's grammar exactly (kept in sync
 * by hand; both sides carry the same example table in their tests). The
 * backend only *validates* a formula at template-save time; this module is
 * the one place that actually *evaluates* one, against a live entity's
 * field values, for display in a filled-in sheet.
 *
 * Grammar (EBNF):
 *   expr           := ternary
 *   ternary        := logic_or ( "?" expr ":" expr )?
 *   logic_or       := logic_and ( "||" logic_and )*
 *   logic_and      := equality ( "&&" equality )*
 *   equality       := comparison ( ("==" | "!=") comparison )*
 *   comparison     := additive ( ("<" | "<=" | ">" | ">=") additive )*
 *   additive       := multiplicative ( ("+" | "-") multiplicative )*
 *   multiplicative := unary ( ("*" | "/" | "%") unary )*
 *   unary          := ("!" | "-")? primary
 *   primary        := NUMBER | IDENT | "(" expr ")" | IDENT "(" (expr ("," expr)*)? ")"
 *
 * `IDENT` not immediately followed by `(` is a field-key reference; `IDENT(`
 * is a call to one of ALLOWED_FUNCTIONS.
 */

export const ALLOWED_FUNCTIONS = new Set([
  "floor",
  "ceil",
  "round",
  "min",
  "max",
  "abs",
]);

const OPERATORS = [
  "<=",
  ">=",
  "==",
  "!=",
  "&&",
  "||",
  "?",
  ":",
  "<",
  ">",
  "+",
  "-",
  "*",
  "/",
  "%",
  "!",
  "(",
  ")",
  ",",
] as const;

export class FormulaSyntaxError extends Error {}

interface Token {
  kind: "number" | "ident" | "op" | "end";
  value: string;
}

function tokenize(source: string): Token[] {
  const tokens: Token[] = [];
  let pos = 0;
  const length = source.length;
  while (pos < length) {
    const ch = source[pos];
    if (/\s/.test(ch)) {
      pos += 1;
      continue;
    }
    if (/[0-9]/.test(ch)) {
      const start = pos;
      while (pos < length && /[0-9]/.test(source[pos])) pos += 1;
      if (source[pos] === "." && /[0-9]/.test(source[pos + 1] ?? "")) {
        pos += 1;
        while (pos < length && /[0-9]/.test(source[pos])) pos += 1;
      }
      tokens.push({ kind: "number", value: source.slice(start, pos) });
      continue;
    }
    if (/[A-Za-z_]/.test(ch)) {
      const start = pos;
      while (pos < length && /[A-Za-z0-9_]/.test(source[pos])) pos += 1;
      tokens.push({ kind: "ident", value: source.slice(start, pos) });
      continue;
    }
    const op = OPERATORS.find((candidate) => source.startsWith(candidate, pos));
    if (op === undefined) {
      throw new FormulaSyntaxError(
        `unexpected character ${JSON.stringify(ch)} at position ${pos}`,
      );
    }
    tokens.push({ kind: "op", value: op });
    pos += op.length;
  }
  tokens.push({ kind: "end", value: "" });
  return tokens;
}

export type FormulaNode =
  | { type: "num"; value: number }
  | { type: "ident"; name: string }
  | { type: "unary"; op: "!" | "-"; arg: FormulaNode }
  | { type: "binary"; op: string; left: FormulaNode; right: FormulaNode }
  | { type: "ternary"; cond: FormulaNode; then: FormulaNode; else: FormulaNode }
  | { type: "call"; name: string; args: FormulaNode[] };

class Parser {
  private pos = 0;
  private readonly tokens: Token[];

  constructor(tokens: Token[]) {
    this.tokens = tokens;
  }

  private peek(): Token {
    return this.tokens[this.pos];
  }

  private advance(): Token {
    return this.tokens[this.pos++];
  }

  private atOp(op: string): boolean {
    const token = this.peek();
    return token.kind === "op" && token.value === op;
  }

  private expectOp(op: string): void {
    if (!this.atOp(op)) {
      throw new FormulaSyntaxError(
        `expected ${JSON.stringify(op)}, found ${JSON.stringify(this.peek().value)}`,
      );
    }
    this.advance();
  }

  parse(): FormulaNode {
    const node = this.expr();
    if (this.peek().kind !== "end") {
      throw new FormulaSyntaxError(
        `unexpected trailing token ${JSON.stringify(this.peek().value)}`,
      );
    }
    return node;
  }

  private expr(): FormulaNode {
    return this.ternary();
  }

  private ternary(): FormulaNode {
    const cond = this.logicOr();
    if (this.atOp("?")) {
      this.advance();
      const then = this.expr();
      this.expectOp(":");
      const elseBranch = this.expr();
      return { type: "ternary", cond, then, else: elseBranch };
    }
    return cond;
  }

  private binaryLevel(next: () => FormulaNode, ops: readonly string[]): FormulaNode {
    let node = next();
    while (this.peek().kind === "op" && ops.includes(this.peek().value)) {
      const op = this.advance().value;
      const right = next();
      node = { type: "binary", op, left: node, right };
    }
    return node;
  }

  private logicOr = (): FormulaNode => this.binaryLevel(this.logicAnd, ["||"]);
  private logicAnd = (): FormulaNode => this.binaryLevel(this.equality, ["&&"]);
  private equality = (): FormulaNode =>
    this.binaryLevel(this.comparison, ["==", "!="]);
  private comparison = (): FormulaNode =>
    this.binaryLevel(this.additive, ["<", "<=", ">", ">="]);
  private additive = (): FormulaNode =>
    this.binaryLevel(this.multiplicative, ["+", "-"]);
  private multiplicative = (): FormulaNode =>
    this.binaryLevel(this.unary, ["*", "/", "%"]);

  private unary = (): FormulaNode => {
    const token = this.peek();
    if (token.kind === "op" && (token.value === "!" || token.value === "-")) {
      this.advance();
      return { type: "unary", op: token.value, arg: this.unary() };
    }
    return this.primary();
  };

  private primary(): FormulaNode {
    const token = this.advance();
    if (token.kind === "number") {
      return { type: "num", value: Number(token.value) };
    }
    if (token.kind === "ident") {
      if (this.atOp("(")) {
        if (!ALLOWED_FUNCTIONS.has(token.value)) {
          throw new FormulaSyntaxError(`unknown function ${JSON.stringify(token.value)}`);
        }
        this.advance();
        const args: FormulaNode[] = [];
        if (!this.atOp(")")) {
          args.push(this.expr());
          while (this.atOp(",")) {
            this.advance();
            args.push(this.expr());
          }
        }
        this.expectOp(")");
        return { type: "call", name: token.value, args };
      }
      return { type: "ident", name: token.value };
    }
    if (token.kind === "op" && token.value === "(") {
      const node = this.expr();
      this.expectOp(")");
      return node;
    }
    throw new FormulaSyntaxError(`unexpected token ${JSON.stringify(token.value)}`);
  }
}

export function parseFormula(formula: string): FormulaNode {
  if (!formula.trim()) {
    throw new FormulaSyntaxError("formula must not be empty");
  }
  return new Parser(tokenize(formula)).parse();
}

export function parseDependencies(formula: string): Set<string> {
  const deps = new Set<string>();
  const walk = (node: FormulaNode): void => {
    switch (node.type) {
      case "ident":
        deps.add(node.name);
        return;
      case "unary":
        walk(node.arg);
        return;
      case "binary":
        walk(node.left);
        walk(node.right);
        return;
      case "ternary":
        walk(node.cond);
        walk(node.then);
        walk(node.else);
        return;
      case "call":
        node.args.forEach(walk);
        return;
      case "num":
        return;
    }
  };
  walk(parseFormula(formula));
  return deps;
}

function toNumber(value: unknown): number {
  if (typeof value === "boolean") return value ? 1 : 0;
  const n = Number(value);
  return Number.isFinite(n) ? n : 0;
}

function toBool(value: unknown): boolean {
  if (typeof value === "boolean") return value;
  return toNumber(value) !== 0;
}

function callFunction(name: string, args: number[]): number {
  switch (name) {
    case "floor":
      return Math.floor(args[0]);
    case "ceil":
      return Math.ceil(args[0]);
    case "round":
      return Math.round(args[0]);
    case "abs":
      return Math.abs(args[0]);
    case "min":
      return Math.min(...args);
    case "max":
      return Math.max(...args);
    default:
      // Unreachable: parseFormula rejects any name outside ALLOWED_FUNCTIONS.
      return 0;
  }
}

function evaluateNode(
  node: FormulaNode,
  values: Record<string, unknown>,
): number | boolean {
  switch (node.type) {
    case "num":
      return node.value;
    case "ident":
      return values[node.name] as number | boolean | undefined ?? 0;
    case "unary": {
      const arg = evaluateNode(node.arg, values);
      return node.op === "!" ? !toBool(arg) : -toNumber(arg);
    }
    case "call":
      return callFunction(
        node.name,
        node.args.map((arg) => toNumber(evaluateNode(arg, values))),
      );
    case "ternary":
      return toBool(evaluateNode(node.cond, values))
        ? evaluateNode(node.then, values)
        : evaluateNode(node.else, values);
    case "binary": {
      if (node.op === "&&") {
        return toBool(evaluateNode(node.left, values)) && toBool(evaluateNode(node.right, values));
      }
      if (node.op === "||") {
        return toBool(evaluateNode(node.left, values)) || toBool(evaluateNode(node.right, values));
      }
      const left = toNumber(evaluateNode(node.left, values));
      const right = toNumber(evaluateNode(node.right, values));
      switch (node.op) {
        case "+":
          return left + right;
        case "-":
          return left - right;
        case "*":
          return left * right;
        case "/":
          return right === 0 ? 0 : left / right;
        case "%":
          return right === 0 ? 0 : left % right;
        case "==":
          return left === right;
        case "!=":
          return left !== right;
        case "<":
          return left < right;
        case "<=":
          return left <= right;
        case ">":
          return left > right;
        case ">=":
          return left >= right;
        default:
          return 0;
      }
    }
  }
}

/** Missing/unresolvable field keys evaluate to 0/false — an entity can be
 * only partially filled in, so a computed widget must degrade, not throw. */
export function evaluateFormula(
  formula: string,
  values: Record<string, unknown>,
): number | boolean {
  return evaluateNode(parseFormula(formula), values);
}

File: backend/src/loregraph/schemas/formula.py
"""Formula grammar for computed sheet widgets — a deliberately small
expression language, not arbitrary Python `eval` (injection risk, unbounded
runtime). This module only *validates*: it checks syntax and extracts the
field keys a formula references. Evaluating a formula against live entity
values is a frontend concern (components/sheet/widgets/formula.ts, which
re-implements this grammar) — the backend has no entity field values to
compute against at template-save time, only field_defs to validate references
against. The two must be changed together: this module's tests cover the
grammar, the evaluator has none of its own, so a token added here and missed
there surfaces only as a sheet showing 0.

Grammar (EBNF)::

    expr           := ternary
    ternary        := logic_or ( "?" expr ":" expr )?
    logic_or       := logic_and ( "||" logic_and )*
    logic_and      := equality ( "&&" equality )*
    equality       := comparison ( ("==" | "!=") comparison )*
    comparison     := additive ( ("<" | "<=" | ">" | ">=") additive )*
    additive       := multiplicative ( ("+" | "-") multiplicative )*
    multiplicative := unary ( ("*" | "/" | "%") unary )*
    unary          := ("!" | "-")? primary
    primary        := NUMBER | IDENT | "(" expr ")" | IDENT "(" (expr ("," expr)*)? ")"

`IDENT` not immediately followed by `(` is a field-key reference; `IDENT(`
is a call to one of ALLOWED_FUNCTIONS.
"""

from collections.abc import Callable
from dataclasses import dataclass

ALLOWED_FUNCTIONS: frozenset[str] = frozenset(
    {"floor", "ceil", "round", "min", "max", "abs"}
)

_OPERATORS = (
    "<=",
    ">=",
    "==",
    "!=",
    "&&",
    "||",
    "?",
    ":",
    "<",
    ">",
    "+",
    "-",
    "*",
    "/",
    "%",
    "!",
    "(",
    ")",
    ",",
)


class FormulaSyntaxError(ValueError):
    """A computed-widget formula does not match the expression grammar."""


@dataclass(frozen=True)
class _Token:
    kind: str  # "number" | "ident" | "op" | "end"
    value: str


def _tokenize(source: str) -> list[_Token]:
    tokens: list[_Token] = []
    pos = 0
    length = len(source)
    while pos < length:
        ch = source[pos]
        if ch.isspace():
            pos += 1
            continue
        if ch.isdigit():
            start = pos
            while pos < length and source[pos].isdigit():
                pos += 1
            if (
                pos < length
                and source[pos] == "."
                and pos + 1 < length
                and source[pos + 1].isdigit()
            ):
                pos += 1
                while pos < length and source[pos].isdigit():
                    pos += 1
            tokens.append(_Token("number", source[start:pos]))
            continue
        if ch.isalpha() or ch == "_":
            start = pos
            while pos < length and (source[pos].isalnum() or source[pos] == "_"):
                pos += 1
            tokens.append(_Token("ident", source[start:pos]))
            continue
        for op in _OPERATORS:
            if source.startswith(op, pos):
                tokens.append(_Token("op", op))
                pos += len(op)
                break
        else:
            raise FormulaSyntaxError(f"unexpected character {ch!r} at position {pos}")
    tokens.append(_Token("end", ""))
    return tokens


class _Parser:
    """Recursive-descent parser that validates syntax without ever computing
    a value — it only walks the grammar and records which identifiers were
    referenced as field keys (as opposed to function-call names)."""

    def __init__(self, tokens: list[_Token]) -> None:
        self._tokens = tokens
        self._pos = 0
        self.dependencies: set[str] = set()

    def _peek(self) -> _Token:
        return self._tokens[self._pos]

    def _advance(self) -> _Token:
        token = self._tokens[self._pos]
        self._pos += 1
        return token

    def _at_op(self, op: str) -> bool:
        token = self._peek()
        return token.kind == "op" and token.value == op

    def _expect_op(self, op: str) -> None:
        if not self._at_op(op):
            raise FormulaSyntaxError(f"expected {op!r}, found {self._peek().value!r}")
        self._advance()

    def parse(self) -> None:
        self._expr()
        if self._peek().kind != "end":
            raise FormulaSyntaxError(
                f"unexpected trailing token {self._peek().value!r}"
            )

    def _expr(self) -> None:
        self._ternary()

    def _ternary(self) -> None:
        self._logic_or()
        if self._at_op("?"):
            self._advance()
            self._expr()
            self._expect_op(":")
            self._expr()

    def _binary_level(
        self, next_level: Callable[[], None], ops: tuple[str, ...]
    ) -> None:
        next_level()
        while self._peek().kind == "op" and self._peek().value in ops:
            self._advance()
            next_level()

    def _logic_or(self) -> None:
        self._binary_level(self._logic_and, ("||",))

    def _logic_and(self) -> None:
        self._binary_level(self._equality, ("&&",))

    def _equality(self) -> None:
        self._binary_level(self._comparison, ("==", "!="))

    def _comparison(self) -> None:
        self._binary_level(self._additive, ("<", "<=", ">", ">="))

    def _additive(self) -> None:
        self._binary_level(self._multiplicative, ("+", "-"))

    def _multiplicative(self) -> None:
        self._binary_level(self._unary, ("*", "/", "%"))

    def _unary(self) -> None:
        token = self._peek()
        if token.kind == "op" and token.value in ("!", "-"):
            self._advance()
            self._unary()
            return
        self._primary()

    def _primary(self) -> None:
        token = self._advance()
        if token.kind == "number":
            return
        if token.kind == "ident":
            if self._at_op("("):
                if token.value not in ALLOWED_FUNCTIONS:
                    raise FormulaSyntaxError(f"unknown function {token.value!r}")
                self._advance()
                if not self._at_op(")"):
                    self._expr()
                    while self._at_op(","):
                        self._advance()
                        self._expr()
                self._expect_op(")")
                return
            self.dependencies.add(token.value)
            return
        if token.kind == "op" and token.value == "(":
            self._expr()
            self._expect_op(")")
            return
        raise FormulaSyntaxError(f"unexpected token {token.value!r}")


def parse_dependencies(formula: str) -> frozenset[str]:
    """Validate a computed-widget formula and return the field keys it
    references. Raises FormulaSyntaxError on invalid syntax, an unbalanced
    expression, or a call to a function outside ALLOWED_FUNCTIONS."""
    if not formula.strip():
        raise FormulaSyntaxError("formula must not be empty")
    parser = _Parser(_tokenize(formula))
    parser.parse()
    return frozenset(parser.dependencies)
